pkg/cloudproduct/gke/gke_test.go

```diff
@@ -77,6 +77,198 @@ func TestSyncPodPortsToGameServer(t *testing.T) {
 	}
 }
 
+func TestValidateGameServer(t *testing.T) {
+	for name, tc := range map[string]struct {
+		gs   *agonesv1.GameServer
+		want []metav1.StatusCause
+	}{
+		"no ports => validated": {
+			gs: &agonesv1.GameServer{},
+		},
+		"good ports => validated": {
+			gs: &agonesv1.GameServer{
+				Spec: agonesv1.GameServerSpec{
+					Ports: []agonesv1.GameServerPort{
+						{
+							Name:          "some-tcpudp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 4321,
+							Protocol:      agonesv1.ProtocolTCPUDP,
+						},
+						{
+							Name:          "awesome-udp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolUDP,
+						},
+						{
+							Name:          "awesome-tcp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolTCP,
+						},
+					},
+				},
+			},
+		},
+		"bad policy => fails validation": {
+			gs: &agonesv1.GameServer{
+				Spec: agonesv1.GameServerSpec{
+					Ports: []agonesv1.GameServerPort{
+						{
+							Name:          "best-tcpudp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 4321,
+							Protocol:      agonesv1.ProtocolTCPUDP,
+						},
+						{
+							Name:          "bad-udp",
+							PortPolicy:    agonesv1.Static,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolUDP,
+						},
+						{
+							Name:          "another-bad-udp",
+							PortPolicy:    agonesv1.Static,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolUDP,
+						},
+					},
+				},
+			},
+			want: []metav1.StatusCause{
+				{
+					Type:    "FieldValueInvalid",
+					Message: "PortPolicy must be Dynamic on GKE Autopilot",
+					Field:   "bad-udp.portPolicy",
+				},
+				{
+					Type:    "FieldValueInvalid",
+					Message: "PortPolicy must be Dynamic on GKE Autopilot",
+					Field:   "another-bad-udp.portPolicy",
+				},
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			causes := (&gkeAutopilot{}).ValidateGameServer(tc.gs)
+			if diff := cmp.Diff(tc.want, causes); diff != "" {
+				t.Errorf("Causes differ (-want +got):\n%s", diff)
+			}
```

Comment:
```suggestion
require.Equal(tc.want, causes)
```

Saves you some time and effort, and provides a diff if different. We use the testify libs pretty much everywhere.

Reply:
Done. Sorry, old habits die hard. :)